core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java

```diff
@@ -1513,6 +1513,34 @@ public static String concatMultiWithNull(String... args) {
   public static String concatMultiWithSeparator(String... args) {
     // the separator arg could be null
     final String sep = args[0] == null ? "" : args[0];
+    return concatMultiWithSeparator(sep, args);
+  }
+
+
+  /** SQL {@code CONCAT_WS(sep[, str | array(str)]+)} function,
+   * return null as null sep. */
+  public static String concatMultiTypeWithSeparator(String sep, Object... args) {
+    if (args.length == 0) {
+      return "";
+    }
+    Object[] argsArray = array(args);
+    List<String> arrayList = new ArrayList<>();
+    arrayList.add(0, sep);
```

Comment:
what happens here?

Reply:
Make `sep` as the first element in the collection, equivalent to `arrayList.add(sep)`.

Reply:
This I understand, but when does this need to happen?

Reply:
The existing methods only support concat for strings. Here extends concat implementations for strings and arrays.